yaspin/core.py

```diff
@@ -533,14 +531,22 @@ def _set_cycle(frames):
 
     @staticmethod
     def _hide_cursor():
-        sys.stdout.write("\033[?25l")
-        sys.stdout.flush()
+        if sys.stdout.isatty():
+            # ANSI Control Sequence DECTCEM 1 does not work in Jupyter
+            sys.stdout.write("\033[?25l")
+            sys.stdout.flush()
 
     @staticmethod
     def _show_cursor():
-        sys.stdout.write("\033[?25h")
-        sys.stdout.flush()
+        if sys.stdout.isatty():
+            # ANSI Control Sequence DECTCEM 2 does not work in Jupyter
+            sys.stdout.write("\033[?25h")
+            sys.stdout.flush()
 
-    @staticmethod
-    def _clear_line():
-        sys.stdout.write("\033[K")
+    def _clear_line(self):
+        if sys.stdout.isatty():
+            # ANSI Control Sequence EL does not work in Jupyter
+            sys.stdout.write("\r\033[K")
+        else:
+            fill = " " * self._cur_line_len
```

Comment:
Do I understand correctly that this is an implementation of EL behavior for Jupyter?

Reply:
Yes, it is. 

`_cur_line_len` keeps track of how many characters have been printed to the current line. `\r` resets the "cursor" to the beginning of the line and `" " * self._cur_line_len` generates enough white spaces to "cover" all previously printed characters.

Reply:
`\033[K` just prints 3 white spaces
`\033[0K` just prints 4 white spaces.

(No idea who decided that behavior... I guess even more misleading than printing `?[K`.)